case-studies/higgs/dataframe/analyzers/APCHiggsTools.cc

```diff
@@ -0,0 +1,117 @@
+#include "APCHiggsTools.h"
+using namespace APCHiggsTools;
+
+ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData>  APCHiggsTools::muon_quality_check(ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> in){
+	ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> result;
+  //at least one muon + and one muon - in each event
+  int n_muon_plus = 0;
+	int n_muon_minus = 0;
+	int n = in.size();
+	for (int i = 0; i < n; ++i) {
+		if (in[i].charge == 1.0){
+			++n_muon_plus;
+		}
+		else if (in[i].charge == -1.0){
+			++n_muon_minus;
+		}
+	}
+	if (n_muon_plus >= 1 && n_muon_minus >= 1){
+		result = in;
+	}
+	return result;
+}
+
+ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> in) {
+   ROOT::VecOps::RVec<float> result;
+	 for (auto & p: in) {
+		 TLorentzVector tlv;
+		 tlv.SetXYZM(p.momentum.x, p.momentum.y, p.momentum.z, p.mass);
+		 result.push_back(tlv.CosTheta());
+	 }
+	 return result;
+}
+
+ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta_miss(ROOT::VecOps::RVec<Float_t>Px, ROOT::VecOps::RVec<Float_t>Py, ROOT::VecOps::RVec<Float_t>Pz, ROOT::VecOps::RVec<Float_t>E) {
```

Comment:
Actually, you do not need this get_cosTheta_miss. You could simply use the previous method, get_cosTheta, and apply it to the "MissingET" RecoParticle (whose name is a bit misleading since the momentum of the latter has a z component, but that is not in your hands). See below in the analysis.py.
I suggest you remove this get_cosTheta_miss, as it may look confusing.

Reply:
Please note that you can overload a function, so you can have both
```ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> in) ```
and 
```ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<Float_t>Px, ROOT::VecOps::RVec<Float_t>Py, ROOT::VecOps::RVec<Float_t>Pz, ROOT::VecOps::RVec<Float_t>E) ```

Reply:
yes of course, although this is not needed here.

Reply:
sure, this is not needed here, sorry for not mentioning it before, just wanted to write it here for future reference

Reply:
> Actually, you do not need this get_cosTheta_miss. You could simply use the previous method, get_cosTheta, and apply it to the "MissingET" RecoParticle (whose name is a bit misleading since the momentum of the latter has a z component, but that is not in your hands). See below in the analysis.py.
> I suggest you remove this get_cosTheta_miss, as it may look confusing.

Hi @EmanuelPerez and @clementhelsens ,
I tried to calculate the costheta missing this way (missing information as an <edm4hep::ReconstructedParticleData> object like the muons), but I failed. Maybe I didn't use the correct method to do so. Yes, of course, this function `ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<Float_t>Px, ROOT::VecOps::RVec<Float_t>Py, ROOT::VecOps::RVec<Float_t>Pz, ROOT::VecOps::RVec<Float_t>E`) is not needed if we could build the standard <edm4hep::ReconstructedParticleData>  with missing informations and then calculate the costheta missing using `APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<edm4hep::ReconstructedParticleData> in)`.

Reply:
Hi Ang, what was your problem / error message ? When I use this:
.Define('missingET_costheta2', 'APCHiggsTools::get_cosTheta( MissingET )')
it works, and produces the same as your
.Define('missingET_costheta', 'APCHiggsTools::get_cosTheta_miss(missingET_px,missingET_py,missingET_pz,missingET_e)')

Reply:
Hi Emmanuel,

I tested it again this morning. 
In analysis.py, I have
`.Alias("MET0", "MissingET#0.index")`
`.Define("MET",  "ReconstructedParticle::get(MET0, ReconstructedParticles)")`
`.Define('MET_costheta', 'APCHiggsTools::get_cosTheta(MET)')`

In finalSel.py, I have 
`"MET_costheta_test":{"name":"MET_costheta","title":"test cos#theta_{missing}","bin":200,"xmin":-1,"xmax":1},`

But the output plot is empty:
[MET_costheta_test_ZH_sel0_stack_lin.pdf](https://github.com/HEP-FCC/FCCeePhysicsPerformance/files/6776140/MET_costheta_test_ZH_sel0_stack_lin.pdf)

When I use 
`ROOT::VecOps::RVec<float> get_cosTheta_miss(ROOT::VecOps::RVec<Float_t>Px, ROOT::VecOps::RVec<Float_t>Py, ROOT::VecOps::RVec<Float_t>Pz, ROOT::VecOps::RVec<Float_t>E);`
the output plot is this one:
[MET_costheta_ZH_sel0_stack_lin.pdf](https://github.com/HEP-FCC/FCCeePhysicsPerformance/files/6776156/MET_costheta_ZH_sel0_stack_lin.pdf)

Reply:
@AngLICERN, you need to do:

```
.Define('missingET_costheta2', 'APCHiggsTools::get_cosTheta( MissingET )')
```

```MissingET``` is already a ```ReconstructedParticle```

Reply:
Yes, Ok, If I build the ReconstructedParticle and get the costheta this way:
`.Alias("MET0", "MissingET#0.index")`
`.Define("MET", "ReconstructedParticle::get(MET0, ReconstructedParticles)")`
`.Define('MET_costheta', 'APCHiggsTools::get_cosTheta(MET)')`

 the output plots is empty. maybe I should use other index instead `MissingET#0.index`?

When I calculate the costheta this way:
`.Define('missingET_px', 'MissingET.momentum.x')`
`.Define('missingET_py', 'MissingET.momentum.y')`
'.Define('missingET_pz', 'MissingET.momentum.z')'
'.Define('missingET_e', 'MissingET.energy')'
'.Define('missingET_costheta', 'APCHiggsTools::get_cosTheta_miss(missingET_px,missingET_py,missingET_pz,missingET_e)')'

The output plot is normal.

Reply:
Ang, could you please try, verbatim, what was suggested:
.Define('missingET_costheta2', 'APCHiggsTools::get_cosTheta( MissingET )')

Your "MET0" above is not the MET at all - it is actually nothing. As said, MissingET is already a ReconstructedParticle. ( Clement, we should document somewhere what the Collection#xx are and how users can figure out what they mean. Obviously Ang got confused when trying to transpose the Muon#0 example to the MET).

Reply:
@EmanuelPerez, the meaning of ```Collection#x``` is again available in the yaml and are the relations:
https://github.com/key4hep/EDM4hep/blob/master/edm4hep.yaml#L361#L368

so you see that ```x``` runs from 0 to 5 in the output file for RecontructedParicles, and we indeed have 6 relations in the yaml.

As we discussed on the mail thread, in RDF we are exposed to the relations because we don't have a nice way (yet) of getting PODIO and RDF to work.

Reply:
Thank you @EmanuelPerez and @clementhelsens , I tested it, and it works well now. Yes, in this case, we don't need `ROOT::VecOps::RVec<float> APCHiggsTools::get_cosTheta(ROOT::VecOps::RVec<Float_t>Px, ROOT::VecOps::RVec<Float_t>Py, ROOT::VecOps::RVec<Float_t>Pz, ROOT::VecOps::RVec<Float_t>E)` anymore.